src/main.rs

```diff
@@ -96,42 +95,59 @@ async fn handle_message(
         }
     };
 
-    // StateChange related to Account
-    let account_changes_future = db_adapters::account_changes::store_account_changes(
-        &pool,
-        &streamer_message.state_changes,
-        &streamer_message.block.header.hash,
-        streamer_message.block.header.timestamp,
-    );
-
-    join!(
-        execution_outcomes_future,
-        accounts_future,
-        access_keys_future,
-        account_changes_future,
-    );
+    if strict_mode {
+        // StateChange related to Account
+        let account_changes_future = db_adapters::account_changes::store_account_changes(
+            &pool,
+            &streamer_message.state_changes,
+            &streamer_message.block.header.hash,
+            streamer_message.block.header.timestamp,
+        );
+
+        join!(
+            execution_outcomes_future,
+            accounts_future,
+            access_keys_future,
+            account_changes_future,
+        );
+    } else {
+        join!(
+            execution_outcomes_future,
+            accounts_future,
+            access_keys_future,
+        );
+    }
 }
 
 async fn listen_blocks(
     stream: mpsc::Receiver<near_indexer::StreamerMessage>,
     pool: Database<PgConnection>,
     concurrency: std::num::NonZeroU16,
-    allow_missing_relation_in_start_blocks: Option<u32>,
+    strict_mode: bool,
+    stop_after_number_of_blocks: Option<u64>,
 ) {
     tracing::info!(target: crate::INDEXER_FOR_EXPLORER, "Stream has started");
-    let strict_mode = allow_missing_relation_in_start_blocks.unwrap_or(0);
-    let mut handle_messages = tokio_stream::wrappers::ReceiverStream::new(stream)
-        .enumerate()
-        .map(|(index, streamer_message)| {
+    let handle_messages =
+        tokio_stream::wrappers::ReceiverStream::new(stream).map(|streamer_message| {
             info!(
                 target: crate::INDEXER_FOR_EXPLORER,
                 "Block height {}", &streamer_message.block.header.height
             );
-            handle_message(&pool, streamer_message, index >= strict_mode as usize)
-        })
-        .buffer_unordered(usize::from(concurrency.get()));
+            handle_message(&pool, streamer_message, strict_mode)
+        });
+    let mut handle_messages = if let Some(stop_after_n_blocks) = stop_after_number_of_blocks {
+        handle_messages
+            .take(stop_after_n_blocks as usize)
```

Comment:
I would either make `stop_after_number_of_blocks` to be `usize` in the first place, or use `usize::try_from().expect()` since `as usize` might cause unexpected issues on non-64-bit platforms down the road.